### ISSUE TITLE
error_log path specified in php.ini is not respected

### ISSUE THREAD

**Issue by metal450** [cytopia/devilbox#577]
1. I have verified that my issue is not already reported on GitHub.

2. I have (done my best to) verify that the issue is not covered in the docs: https://devilbox.readthedocs.io

### ISSUE TYPE

 - Bug Report

### Checklist
* [X] `.env` file is attached (renamed to env.txt so Github would accept it)
* [ ] `docker-compose.override.yml` is attached (if it exists)
* [X] Custom configs from `cfg/` dir are attached (if customized)
* [ ] `docker-compose logs` and `log/` output is added
* [X] I've looked through the docs: https://devilbox.readthedocs.io/en/latest/
* [X] I've looked through existing issues: https://github.com/cytopia/devilbox/issues
* [X] I've read troubleshooting: https://devilbox.readthedocs.io/en/latest/support/troubleshooting.html

### OS / ENVIRONMENT
1. Host operating system and version: Windows 10 1607
2. (Windows) Native Docker or Docker Toolbox: Docker Toolbox
3. Docker version: 18.09.3
4. Docker Compose version: 1.23.1
5. (Linux) Is SELinux enabled?: n/a
6. What git commit hash are you on?: 4be9018c2bbe19035f759729085010e5995b181d

### SUMMARY
php’s error_log path cannot be configured via php.ini.  Other options configured in php.ini (i.e. upload_max_filesize, allow_url_include, xdebug.xxx, etc) are reflected in phpinfo(), but no matter what value is set for error_log, PHP just ignores it & uses /var/log/php/php-fpm.error.

I know that the documentation (https://devilbox.readthedocs.io/en/latest/intermediate/read-log-files.html) describes that this is where the logs will go.  I would like to customize a different path (i.e. ${DOCUMENT_ROOT}/error_log.log). 

How can DevilBox be made to respect the value set for error_log in php.ini, like in XAMPP & other development environments?

### STEPS TO REPRODUCE
1) Add a custom php.ini with i.e.: error_log = /var/log/php_test.log
2) Start DevilBox
3) View phpinfo
4) See that error_log shows /var/log/php/php-fpm.error, not /var/log/php_test.log as specified in php.ini

### EXPECTED BEHAVIOUR
Respect the error_log setting from php.ini, like other custom settings in php.ini

### ACTUAL BEHAVIOUR
See above

#### Start command
$ docker-compose up -d

#### File and user permissions (Linux & MacOS)
n/a

[env.txt](https://github.com/cytopia/devilbox/files/3155705/env.txt)
[phpini.txt](https://github.com/cytopia/devilbox/files/3155718/phpini.txt)

**Comment by science695** [comment]
Check the fpm config settings, there are fpm config settings that are set to override the php.ini settings

**Comment by metal450** [comment]
*If you're referring to cfg\php-fpm-5.6\devilbox-fpm.conf-default, I don't see anything in there relevant to logging.
*At https://devilbox.readthedocs.io/en/latest/configuration-files/php-fpm-conf.html, I found reference to "php_admin_flag," but that doesn't seem relevant to PHP 5.6 (& that string is also not in the cfg\php-ftp-5.6 files).
*I did a recursive search of the devilbox dir for "fpm" and don't see anything else.

Could you please be specific where to change the "fpm confg settings," and what to change, to get it to respect a different log dir?

**Comment by metal450** [comment]
Would really appreciate some feedback.  I've been trying to figure this out for over a month (I also posted [on the forum](https://devilbox.discourse.group/t/php-ini-error-log-not-respected/135), but haven't received any replies there either). :/

**Comment by science695** [comment]
If you look up php.ini settings and php fpm settings, you should be able to add them into a file like the devilbox-fpm.conf-default (remove the -default to enable it), or devilbox-php.ini-default.

If the php ini doesn't update it, look up that same setting in php fpm because fpm can override php.ini settings.

I had to discover this by digging around inside the php container, but thanks to devilbox now having more configuration file mounts you can set these settings without going too much under the hood.

**Comment by metal450** [comment]
The problem with adding them to devilbox-fpm.conf is that the DOCUMENT_ROOT environment variables don't seem to work there.  The environment var does work in devilbox-php.ini...but as mentioned above, its error_log value isn't respected.  So there seems to be no way to make the error_log appear in the appropriate document root.

The real issue is, why is devilbox overriding/ignoring the standard php.ini value (there's nothing apparent in the default devilbox-fpm, which can be removed, to get php.ini to behave as it's supposed to.)

**Comment by science695** [comment]
Have you checked the ErrorLog settings in the vhost templates?

I confused your issue with my issue about error reporting. The vhost templates can be set for each project custom. You can copy the global default template from devilbox/cfg/vhost-gen/ (there are examples for the different front end servers there).

Once you have customized your template, you can leave it in this folder renaming it to end in .yml, or take it to a specific projects ./.devilbox/ folder.

Hope this helps

**Comment by metal450** [comment]
I hadn't been using vhost templates at all - just created project directories, per https://devilbox.readthedocs.io/en/latest/getting-started/create-your-first-project.html#step-2-create-a-project-directory.

If I'm understanding correctly, your suggestion is to create a template that applies to all vhosts, and override ERROR_LOG in here?  But this still seems like it's a "global" log setting - not just for PHP logs.  That's part of the issue: by default, Devilbox meshes together all the FPM logging into the same file as PHP errors - which makes it extremely hard to notice when there are actual code issues that should be addressed in your php scripts.  Unless I'm misunderstanding, changing ERROR_LOG in the yaml file would just move the one, global, merged log file.  It doesn't really address the fact that it's breaking the php.ini error_log setting, preventing "splitting up" logs so PHP error appear in the docroot, but leaves the fpm logging elsewhere (like how normal servers behave).

**Comment by science695** [comment]
With Virtualhost templates, you have the option of putting it in the main folder for a global template, or adding it to each project in its respective .devilbox/apache24.yml

Then you could copy the template to each location and manually override the log file for each one.

**Comment by metal450** [comment]
Still trying to figure out how this works, but no avail.

*To start, I just renamed apache24.yml-example-vhost to apache24.yml - that works (of course)
*If I add the following (hardcoded path, just as a test) under the php_fpm section, the server fails to start: _php_value  error_log  /shared/httpd/myproj/htdocs/error_log.log_
*Likewise, it fails to start if I add it under the docroot section.
*Likewise, it fails to start if I add it in the <VirtualHost> tag of the main vhost section.
*Likewise, it fails to start if I change php_value to php_admin_value

....So still, nothing seems to work... :/

**Comment by metal450** [comment]
Not stale - the developer has still not addressed this issue.

**Comment by metal450** [comment]
Still not stale.  Six months since it was reported...but no solution = not stale.

**Comment by cytopia** [comment]
@metal450 this is not going to work as the user-supplied config is read (alphabetically) before the hard-coded logging directive:

![Screenshot 2019-12-31 17-10-19  selection](https://user-images.githubusercontent.com/12533999/71626996-4c45a700-2bf0-11ea-94a2-20d2f52b7eda.png)

* `yyy-devilbox-user-runtime-zzz.ini` is my custom specified php ini `zzz.ini` in `cfg/php-ini-X.Y/zzz.ini`
* `zzz-entrypoint-logfiles.ini` is added by Devilbox entrypoint script (and contains the logfile definition).

The logfiles directive is hardcoded and forced for a good reason. The entrypoint script needs to know the location so it will be able to adjust the ownership of the file specified via `NEW_UID` and `NEW_GID` in the `.env` file. Otherwise it would be created with an ownership that might not be writable from your host system.

So in short, this is not a bug, but a desired feature and unlikely to change as it would bring too many disadvantages.

### How to do what you need?

There is a workaround you can easily do yourself. Use the custom/autostart script feature and create a php ini file in `/usr/local/etc/php/conf.d` inside the PHP container with a name that alphabetically comes after `zzz-entrypoint-logfiles.ini` and you're good to go.

* https://devilbox.readthedocs.io/en/latest/autostart/custom-scripts-per-php-version.html
* https://devilbox.readthedocs.io/en/latest/autostart/custom-scripts-globally.html
